### PR TITLE
feat(weave): Add improvements to Dataset

### PR DESCRIPTION
**Pull request by andrewtruong**
Resolves:
- https://wandb.atlassian.net/browse/WB-19384

This PR adds limited mutation support to `Dataset` and "ref-get-Datasets", including new `append`, `pop`, and comparison methods, allowing users to save and pick up progress from existing datasets.

For "ref-get-Datasets", special `append` and `pop` methods are patched on to the WeaveObject to facilitate similar behaviour without resorting to ops that would otherwise be noisy.

### CODE REVIEW

File: weave/flow/dataset.py [hunk 1]
```diff
@@ -64,3 +64,21 @@ def convert_to_table(cls, rows: Any) -> weave.Table:
...
+
+    def __eq__(self, other: Any) -> bool:
+        return self.rows == other
+
+    def __iadd__(self, rows: list[dict]) -> None:
+        if not all(isinstance(row, dict) for row in rows):
+            raise ValueError("Can only append dicts to Dataset")
+        for row in rows:
+            self.rows.append(row)
+        return self
```
Comment (andrewtruong) [inline weave/flow/dataset.py:76]: `__iadd__` support for `ds += [...]`

This is experimental, please see the other `__iadd__` on `WeaveObject` to see if we want to support this pattern.

From a user perspective it does feel more ergonomic, but these classes can quickly get branchy with a bunch of different cases like we have in `make_trace_object`.  I find the branching harder to read, but maybe it's ok?

File: weave/tests/test_dataset.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+    ds3 = ref2.get()
+    assert ds3 == [
+        {"a": 1, "b": 2},
+        {"a": 3, "b": 4},
+        {"a": 5, "b": 6},
+    ]
+
+
+@pytest.mark.xfail(reason="Table API does not properly handle objects yet")
```
Comment (andrewtruong) [inline weave/tests/test_dataset.py:27]: TODO in a follow up PR.  This is related to Thomas's `Point(x,y)` dataset issue:
https://wandb.atlassian.net/browse/WB-20236

File: weave/trace/vals.py [hunk 3]
```diff
@@ -220,8 +220,24 @@ def __repr__(self) -> str:
...
         return self._val == other
 
+    def __iadd__(self, other: Any) -> Any:
+        if getattr(self._val, "_class_name", None) == "Dataset":
+            rows = other
+            if not all(isinstance(row, dict) for row in rows):
+                raise ValueError("Can only append dicts to Dataset")
+            for row in rows:
+                self._val.rows.append(row)
+            return self
```
Comment (andrewtruong) [inline weave/trace/vals.py:234]: This branching is ok, but could get noisy!

File: weave/trace/vals.py [hunk 4]
```diff
@@ -536,9 +552,22 @@ def make_trace_obj(
 
     if not isinstance(val, Traceable):
         if isinstance(val, ObjectRecord):
-            return WeaveObject(
-                val, ref=new_ref, server=server, root=root, parent=parent
-            )
+            res = WeaveObject(val, ref=new_ref, server=server, root=root, parent=parent)
```
Comment (shawnlewis) [inline weave/trace/vals.py:555]: Why not initialize a Dataset here instead?

**Review by shawnlewis — CHANGES_REQUESTED** [review]

File: weave/trace/vals.py (hunk 4 above)
Comment (andrewtruong) [reply weave/trace/vals.py:557]: good point -- we're special casing our known objects here so may as well just return the actual `Dataset`

**Pull request closed without merging**